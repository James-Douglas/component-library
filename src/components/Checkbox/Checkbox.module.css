label {
  @apply flex;
}
.manor-checkbox {
  @apply bg-white border border-dark-blue cursor-pointer mb-8 w-24 h-24;
  border-radius: 0.2rem;
}
.manor-checkbox:hover {
  @apply border-light-blue;
}
.manor-checkbox-input:invalid + label > .manor-checkbox{
  @apply border-invalid text-invalid fill-current;
}
.manor-checkbox-input:disabled + label > .manor-checkbox{
  @apply cursor-not-allowed;
}
.manor-checkbox-input:disabled + label > .manor-checkbox{
  @apply bg-grey-light border-none opacity-50;
}
.manor-checkbox-input:checked + label > .manor-checkbox {
  @apply bg-dark-blue border-none text-white fill-current;
}
.manor-checkbox-input:checked + label > .manor-checkbox.inverted {
  @apply bg-white border-none text-dark-blue fill-current;
}
.disabled {
  @apply pointer-events-none;
}