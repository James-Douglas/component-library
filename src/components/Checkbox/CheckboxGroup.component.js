import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { hasTooltipContent, getScreenReaderLabel } from 'utils/form';
import Row from '../Grid/Row/Row.component';
import Column from '../Grid/Column/Column.component';
import { InlineTooltip, tooltipPropTypes } from '../Tooltip/Tooltip.component';
import Label from '../Label/Label.component';
import useIsDesktop from '../../hooks/useIsDesktop';
import FieldValidation from '../FieldValidation/FieldValidation.component';

const StyledContainer = styled.div`
  width: 100%;
`;

const StyledRow = styled(Row)`
  margin-bottom: ${(props) => props.theme.spacing[16]};
`;

export const generateGroup = (colSize, children, callback) => {
  if (children) {
    return children.map((child) => {
      const component = React.cloneElement(child, { handleChange: callback });
      return (
        <Column cols={colSize} key={`key-${child.props.id}`}>
          {component}
        </Column>
      );
    });
  }
  return null;
};

const CheckboxGroup = ({
  label,
  tooltip,
  forceFullWidth,
  validationMessage,
  groupId,
  colSize,
  handleChange,
  children,
}) => {
  const [selectedCheckboxes, setSelectedCheckboxes] = useState([]);
  const desktop = useIsDesktop(false);

  const handleCheckboxClick = ({ id, value }) => {
    const exists = selectedCheckboxes.find((checkbox) => checkbox.id === id);
    if (value && !exists) {
      const newSelectedCheckboxes = [...selectedCheckboxes, { id, value }];
      setSelectedCheckboxes(newSelectedCheckboxes);
    } else if (!value && exists) {
      const newSelectedCheckboxes = selectedCheckboxes.filter((selectedCheckbox) => selectedCheckbox.id !== id);
      setSelectedCheckboxes(newSelectedCheckboxes);
    }
  };

  useEffect(() => {
    handleChange(selectedCheckboxes);
  }, [handleChange, selectedCheckboxes]);

  return (
    <>
      <Label forId={groupId} text={label} tooltip={tooltip} fullWidth={forceFullWidth} />
      <StyledRow>
        <Column cols={desktop && !forceFullWidth ? '10' : '12'}>
          <StyledContainer id={groupId}>
            <Row>
              {generateGroup(colSize, children, handleCheckboxClick)}
            </Row>
          </StyledContainer>
          <FieldValidation message={validationMessage} />
        </Column>
        {desktop && hasTooltipContent(tooltip)
        && (
          <Column cols={2}>
            <InlineTooltip
              title={tooltip.title}
              body={tooltip.body}
              screenReaderLabel={getScreenReaderLabel(tooltip.screenReaderLabel, label)}
              justifyEnd={tooltip.justifyEnd}
            />
          </Column>
        )}
      </StyledRow>
    </>
  );
};

CheckboxGroup.propTypes = {
  /**
   * The checkbox ID. This is required, as it informs the label and the value of the checkbox.
   */
  groupId: PropTypes.string.isRequired,
  /**
   * Label for the ToggleGroup.
   */
  label: PropTypes.string,
  /**
   * Tooltip object (see Tooltip documentation)
   */
  tooltip: PropTypes.shape(tooltipPropTypes),
  /**
   * Forces the ToggleGroup to expand to 12 columns (default true for ToggleGroup)
   */
  forceFullWidth: PropTypes.bool,
  /**
   * Displays given validation message and invalid styles on the component when provided.
   */
  validationMessage: PropTypes.string,
  /**
   * Defines the sizing of the columns to wrap the checkboxes in.
   */
  colSize: PropTypes.string,
  /**
   * Called whenever a checkbox is selected or deselected. Is called with an array of objects representing the currently
   selected checkboxes (in the form `{id, value}`)
   */
  handleChange: PropTypes.func,
  /**
   * The child Checkbox components to render in the CheckboxGroup
   */
  children: (props, propname, componentName) => componentName === 'Checkbox',
};

CheckboxGroup.defaultProps = {
  label: '',
  tooltip: {},
  forceFullWidth: false,
  validationMessage: null,
  colSize: '6',
  handleChange: () => { },
  children: [],
};

export default CheckboxGroup;
