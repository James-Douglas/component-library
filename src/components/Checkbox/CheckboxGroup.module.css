.checkbox-content {
  @apply ml-20 cursor-pointer;
}
.disabled {
  @apply pointer-events-none;
}