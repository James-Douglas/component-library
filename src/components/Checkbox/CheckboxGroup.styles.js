import styled from 'styled-components';
import Row from '../Grid/Row/Row.component';

export const StyledContainer = styled.div`
  width: 100%;
`;

export const StyledInnerRow = styled(Row)`
  margin: 0;
`;
