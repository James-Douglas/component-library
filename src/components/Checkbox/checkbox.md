# Checkbox component<br/>
## Usage
~~~js
<Checkbox
	id={'example-id'}
	disabled={false}
	checked={false}
	invertColour={false}
	icon={'check'}
    handleClick={func}
    handleKeyUp={func}
    children={<p>checkbox content</p>}
/>
~~~

## Props
Checkbox is a custom checkbox, styled with an icon. This is the default, singular variety.

1. `id` *(required) string*
    * The checkbox ID. This is required, as it informs the label and the value of the checkbox.
2. `disabled` *(optional) boolean*
    * Defines if the checkbox is disabled or not. Defaults to false.
3. `checked` *(optional) boolean*
    * Defines the checkbox checked state. Defaults to false.
4. `invertColour` *(optional) boolean*
    * Changes the color of the checked state - blue bg, white tick, or white bg, blue tick.
6. `icon` *(optional) string*
    * Defines the icon needed for the checkbox. Defaults to 'check'.
6. `handleClick` *(optional) func*
    * Supplies the label with a custom click function.
6. `handleKeyUp` *(optional) string*
    * Supplies the label with a custom keyUp function.
6. `children` *(optional) string*
    * Defines the associated content for the checkbox, used by the wrapper component, `<CheckboxGroup/>`. Not required for the singular checkbox.


