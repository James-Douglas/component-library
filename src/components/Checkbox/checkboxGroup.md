# CheckboxGroup component<br/>
## Usage
~~~js
<CheckboxGroup groupId={'example-id'} colSize={'6'} handleClick={func} handleKeyUp={func}>
    <Checkbox id="A-1" icon="check"><p>A-1 check</p></Checkbox>
    <Checkbox id="A-2" icon="check"><p>A-2 check</p></Checkbox>
    <Checkbox id="A-3" icon="check"><p>A-3 check</p></Checkbox>
    <Checkbox id="A-4" icon="check"><p>A-4 check</p></Checkbox>
    <Checkbox id="A-5" icon="check"><p>A-5 check</p></Checkbox>
    <Checkbox id="A-6" icon="check"><p>A-6 check</p></Checkbox>
</CheckboxGroup>
~~~

## Props
Checkbox is a custom checkbox, styled with an icon. This is the default, singular variety.

1. `groupId` *(required) string*
    * The checkbox ID. This is required, as it informs the label and the value of the checkbox.
2. `colSize` *(optional) boolean*
    * Defines the sizing of the columns to wrap the checkboxes in.
6. `handleClick` *(optional) func*
    * Supplies a custom click function to the div that wraps the checkboxes.
6. `handleKeyUp` *(optional) string*
    * Supplies a custom click function to the div that wraps the checkboxes.
6. `children` *(optional) component*
    * The checkboxes that are needed for the group


