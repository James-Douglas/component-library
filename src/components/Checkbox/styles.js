import css from 'styled-jsx/css';

export default css`
label {
  @apply flex cursor-pointer;
}
.manor-checkbox {
  @apply bg-white border border-dark-blue mb-8 w-24 h-24 pointer-events-none;
  border-radius: 0.2rem;
}
.manor-checkbox:hover {
  @apply border-light-blue;
}
.manor-checkbox-input {
  opacity: 0;
  width: 0;
  height: 0;
}
.manor-checkbox-input:invalid + label > .manor-checkbox{
  @apply border-invalid text-invalid fill-current;
}
.manor-checkbox-input:disabled + label > .manor-checkbox{
  @apply cursor-not-allowed;
}
.manor-checkbox-input:disabled + label > .manor-checkbox{
  @apply bg-grey-light border-none opacity-50;
}
.manor-checkbox-input:checked + label > .manor-checkbox {
  @apply bg-dark-blue border-none text-white fill-current;
}
.manor-checkbox-input:checked + label > .manor-checkbox.inverted {
  @apply bg-white border-none text-dark-blue fill-current;
}
.checkbox-content {
  @apply ml-20 pointer-events-none;
}
.manor-checkbox-input:focus + label > .manor-checkbox {
  box-shadow: 0 0 0 1px rgba(0, 123, 255, .5);
}
.disabled {
  @apply pointer-events-none;
}
`;
