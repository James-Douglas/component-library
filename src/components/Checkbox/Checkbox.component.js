import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import useToggleState from '../../hooks/useToggleState';
import Icon from '../Icon/Icon.component';
import styles from './styles';

export const renderIcon = (icon, value) => {
  if (value) {
    return (
      <Icon name={icon} size={2} />
    );
  }
  return null;
};

const renderContent = (children) => {
  if (children) {
    return (
      <>
        <style jsx>{styles}</style>
        <div className="checkbox-content no-margin">
          {children}
        </div>
      </>
    );
  }
  return null;
};

const Checkbox = ({
  id,
  icon,
  disabled,
  invalid,
  invertColour,
  handleChange,
  isSelected,
  children,
}) => {
  const [value, toggle] = useToggleState(isSelected);

  const toggleEventHandler = () => {
    toggle(value);
  };

  useEffect(() => {
    handleChange({ id, value });
  }, [handleChange, id, value]);

  return (
    <>
      <style jsx>{styles}</style>
      <div className="checkbox-wrap">
        <input
          id={id}
          name={id}
          type="checkbox"
          disabled={disabled}
          onChange={toggleEventHandler}
          checked={value}
          className="manor-checkbox-input"
        />

        <label
          htmlFor={id}
          className={`
            ${disabled ? 'disabled' : ''}
            ${children ? '' : 'ap alignment'}
          `}
        >
          <div className={`
            manor-checkbox 
            ${invertColour ? 'inverted' : ''}
            ${invalid ? 'manor-checkbox-invalid' : ''}
          `}
          >
            {renderIcon(icon, value)}
          </div>

          {renderContent(children)}

        </label>
      </div>
    </>
  );
};

Checkbox.propTypes = {
  id: PropTypes.string.isRequired,
  icon: PropTypes.string,
  disabled: PropTypes.bool,
  isSelected: PropTypes.bool,
  invalid: PropTypes.bool,
  invertColour: PropTypes.bool,
  handleChange: PropTypes.func,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]),
};

Checkbox.defaultProps = {
  icon: 'check',
  disabled: false,
  isSelected: false,
  invalid: false,
  invertColour: false,
  handleChange: () => { },
  children: null,
};

export default Checkbox;
