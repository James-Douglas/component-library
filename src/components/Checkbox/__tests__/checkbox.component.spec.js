import React from 'react';
import { faTimes } from '@fortawesome/pro-regular-svg-icons/faTimes';
import { render, fireEvent } from '../../../testUtils';
import Checkbox, { renderIcon } from '../Checkbox.component';
import theme from '../../../themes/ctm.theme';

describe('renderIcon()', () => {
  // eslint-disable-next-line react/prop-types
  const IconContainer = ({ icon, toggle }) => (
    <>
      {renderIcon(icon, toggle)}
    </>
  );

  it('does not render an icon if toggle condition is not met', () => {
    const { container } = render(<IconContainer icon={faTimes} toggle={false} />);
    expect(container).toBeEmpty();
  });

  it('renders an icon when params are set (icon name and toggle is true)', () => {
    const { container } = render(<IconContainer icon={faTimes} toggle />);
    const svg = container.querySelector('svg');
    expect(svg).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });
});

describe('Checkbox', () => {
  it('renders with minimal props', () => {
    const { container } = render(<Checkbox id="test-id" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders with props', () => {
    const mockTestClick = jest.fn();
    const { container } = render(
      <Checkbox
        id="test-id"
        icon={faTimes}
        disabled={false}
        invertColour
        handleChange={mockTestClick}
      >
        <p>child content</p>
      </Checkbox>,
    );

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');
    const labelStyle = label.firstChild;

    fireEvent.click(label, { button: 0 });

    expect(mockTestClick).toHaveBeenCalled();
    expect(labelStyle).toHaveStyle(`background: ${theme.checkbox.background}`);
    expect(checkbox.getAttribute('disabled')).toBe(null);
    expect(container.querySelector('#child-content')).toBeDefined();
    expect(container.getElementsByTagName('svg')).toBeDefined();
    expect(container.querySelector('.fa-check')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders an invalid state', () => {
    const { container } = render(
      <Checkbox
        id="test-id"
        validationMessage="error"
      >
        <p>child content</p>
      </Checkbox>,
    );

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');
    const labelStyle = label.firstChild;
    fireEvent.click(label, { button: 0 });

    expect(labelStyle).toHaveStyle(`border: ${theme.checkbox.invalid}`);
    expect(checkbox.getAttribute('disabled')).toBe(null);
    expect(container.querySelector('#child-content')).toBeDefined();
    expect(container.getElementsByTagName('svg')).toBeDefined();
  });

  it('accepts a prefill value', () => {
    const { container } = render(
      <Checkbox
        id="test-id"
        isSelected
      />,
    );

    const checkbox = container.querySelector('#test-id');
    expect(checkbox.checked).toBe(true);
  });

  it('checks on click', () => {
    const { container } = render(<Checkbox id="test-id" disabled={false} invertColour />);

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');

    fireEvent.click(label, { button: 0 });

    expect(checkbox.checked).toBe(true);
  });

  it('click is disabled when disabled prop is supplied', () => {
    const { container } = render(<Checkbox id="test-id" disabled />);

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');

    fireEvent.click(label, { button: 0 });

    expect(checkbox.checked).toBe(false);
  });

  it('calls focus handler on focus', () => {
    const handleFocus = jest.fn();
    const { container } = render(<Checkbox id="test-id" handleFocus={handleFocus} />);
    const input = container.querySelector('input');
    input.focus();
    expect(handleFocus).toHaveBeenCalled();
  });

  it('calls blur handler on blur', () => {
    const handleBlur = jest.fn();
    const { container } = render(<Checkbox id="test-id" handleBlur={handleBlur} />);
    const input = container.querySelector('input');
    input.focus();
    input.blur();
    expect(handleBlur).toHaveBeenCalled();
  });
});
