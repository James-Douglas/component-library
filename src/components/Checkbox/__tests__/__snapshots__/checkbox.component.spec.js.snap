// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox.component renders with minimal props 1`] = `"<div class=\\"sc-AykKD kORHBz\\"><input id=\\"test-id\\" name=\\"test-id\\" type=\\"checkbox\\" class=\\"sc-AykKC jVorYY\\"><label for=\\"test-id\\" class=\\"sc-AykKE fnCsWZ\\"><div class=\\"sc-AykKF kbzaRi\\"></div></label></div>"`;

exports[`Checkbox.component renders with props 1`] = `"<div class=\\"sc-AykKD kORHBz\\"><input id=\\"test-id\\" name=\\"test-id\\" type=\\"checkbox\\" class=\\"sc-AykKC jVorYY\\"><label for=\\"test-id\\" class=\\"sc-AykKE cPvVnI\\"><div class=\\"sc-AykKF gvEQTf\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"far\\" data-icon=\\"times\\" class=\\"svg-inline--fa fa-times fa-w-10 fa-1x \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 320 512\\"><path fill=\\"currentColor\\" d=\\"M207.6 256l107.72-107.72c6.23-6.23 6.23-16.34 0-22.58l-25.03-25.03c-6.23-6.23-16.34-6.23-22.58 0L160 208.4 52.28 100.68c-6.23-6.23-16.34-6.23-22.58 0L4.68 125.7c-6.23 6.23-6.23 16.34 0 22.58L112.4 256 4.68 363.72c-6.23 6.23-6.23 16.34 0 22.58l25.03 25.03c6.23 6.23 16.34 6.23 22.58 0L160 303.6l107.72 107.72c6.23 6.23 16.34 6.23 22.58 0l25.03-25.03c6.23-6.23 6.23-16.34 0-22.58L207.6 256z\\"></path></svg></div><div class=\\"sc-AykKG zMzrB checkbox-content\\"><p>child content</p></div></label></div>"`;

exports[`renderIcon() renders an icon when params are set (icon name and toggle is true) 1`] = `"<svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"far\\" data-icon=\\"times\\" class=\\"svg-inline--fa fa-times fa-w-10 fa-1x \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 320 512\\"><path fill=\\"currentColor\\" d=\\"M207.6 256l107.72-107.72c6.23-6.23 6.23-16.34 0-22.58l-25.03-25.03c-6.23-6.23-16.34-6.23-22.58 0L160 208.4 52.28 100.68c-6.23-6.23-16.34-6.23-22.58 0L4.68 125.7c-6.23 6.23-6.23 16.34 0 22.58L112.4 256 4.68 363.72c-6.23 6.23-6.23 16.34 0 22.58l25.03 25.03c6.23 6.23 16.34 6.23 22.58 0L160 303.6l107.72 107.72c6.23 6.23 16.34 6.23 22.58 0l25.03-25.03c6.23-6.23 6.23-16.34 0-22.58L207.6 256z\\"></path></svg>"`;
